### ISSUE TITLE
Mounting home in singularity containers is bad for security, reproducibility and creates race conditions

### ISSUE THREAD

**Issue by rhpvorderman** [galaxyproject/galaxy#18620]
**Galaxy Version and/or server at which you observed the bug**
Galaxy Version: 24.1

### Expected behaviour
Singularity mounts no home directory as that can affect reproducibility (settings from home influencing tool behavior).

### Actual behaviour
/home/galaxy is mounted by default in singularity images. In fact the `--home` flag pointing to it is passed to the singularity command, making it execute the command in /home/galaxy. 

This is rather undesirable.

- Security-wise, as this allows access to the ssh keys that galaxy uses on the cluster, if a tool has a vulnerability that allows arbitrary code execution to read such keys. 
- For reproducibility: jobs should not be affected by any settings in any home dir, but be consistently the same on any galaxy instance.
- For concurrency. The user home is mounted on all cluster nodes, meaning that path conflicts (race conditions) occur in the home directory when multiple jobs run simultaneously. 

### How to reproduce
Using a simple job configuration:
```xml
<?xml version="1.0"?>
<!-- A sample job config that explicitly configures job running the way it is
     configured by default (if there is no explicit config). -->
<job_conf>
    <plugins>
        <plugin id="local" type="runner" load="galaxy.jobs.runners.local:LocalJobRunner" workers="4"/>
    </plugins>
    <destinations default="singularity">
        <destination id="local_destination" runner="local"/>
        <destination id="singularity" runner="local">
            <param id="tmp_dir">True</param>
            <param id="singularity_enabled">true</param>
            <param id="singularity_run_extra_arguments">--contain --ipc --pid</param>
            <!-- Ensuring a consistent locale environment is good for reproducibility. -->
            <env id="LC_ALL">C</env>
            <!-- The cache directory holds the docker containers that get converted. -->
            <env id="SINGULARITY_CACHEDIR">/tmp/singularity</env>
        </destination>
    </destinations>
</job_conf>
```

Please note that using the `--contain` flag, which explicitly does not allow singularity to mount /home, does not work here, because galaxy itself is mounting /home.

In the resulting script the following command line is produced:
```
cd working; SINGULARITYENV_GALAXY_SLOTS="$GALAXY_SLOTS" SINGULARITYENV_GALAXY_MEMORY_MB="$GALAXY_MEMORY_MB" SINGULARITYENV_GALAXY_MEMORY_MB_PER_SLOT="$GALAXY_MEMORY_MB_PER_SLOT" SINGULARITYENV_HOME="$HOME" SINGULARITYENV__GALAXY_JOB_HOME_DIR="$_GALAXY_JOB_HOME_DIR" SINGULARITYENV__GALAXY_JOB_TMP_DIR="$_GALAXY_JOB_TMP_DIR" SINGULARITYENV_TMPDIR="$TMPDIR" SINGULARITYENV_TMP="$TMP" SINGULARITYENV_TEMP="$TEMP" singularity -s exec --cleanenv --no-mount tmp -B "$_GALAXY_JOB_TMP_DIR:$_GALAXY_JOB_TMP_DIR" -B "$TMPDIR:$TMPDIR" -B "$TMP:$TMP" -B "$TEMP:$TEMP" -B /cluster/galaxy/server:/cluster/galaxy/server -B /cluster/galaxy/data/shed_tools/toolshed.g2.bx.psu.edu/repos/devteam/fastqc/5ec9f6bceaee/fastqc:/cluster/galaxy/data/shed_tools/toolshed.g2.bx.psu.edu/repos/devteam/fastqc/5ec9f6bceaee/fastqc -B /cluster/galaxy/data/jobs/000/6:/cluster/galaxy/data/jobs/000/6 -B /cluster/galaxy/data/jobs/000/6/outputs:/cluster/galaxy/data/jobs/000/6/outputs -B /cluster/galaxy/data/jobs/000/6/configs:/cluster/galaxy/data/jobs/000/6/configs -B /cluster/galaxy/data/jobs/000/6/working:/cluster/galaxy/data/jobs/000/6/working -B /cluster/galaxy/data/datasets:/cluster/galaxy/data/datasets -B /cluster/galaxy/data/tool_data:/cluster/galaxy/data/tool_data --home $HOME:$HOME --contain --ipc --pid docker://quay.io/biocontainers/fastqc:0.12.1--hdfd78af_0 /bin/bash /cluster/galaxy/data/jobs/000/6/tool_script.sh > '../outputs/tool_stdout' 2> '../outputs/tool_stderr'; return_code=$?; echo $return_code > /cluster/galaxy/data/jobs/000/6/galaxy_6.ec; 

```
Please note the `--home $HOME:$HOME` in the resulting script. 

### Temporary mitigation
```
            <param id="singularity_run_extra_arguments">--contain --ipc --pid --pwd $PWD</param>
```
Using the `--pwd $PWD` argument ensures that at least the tool script does not run in home. This mitigates the concurrency problems.

### The problematic code:

https://github.com/galaxyproject/galaxy/blob/2e7459cc319789ba9a2612463eb9c87689e29c48/lib/galaxy/tool_util/deps/singularity_util.py#L61-L105

**Comment by rhpvorderman** [comment]
I can work on this tomorrow. I wonder if other singularity users do depend on this home mounting mechanic.

**Comment by bernt-matthias** [comment]
Probably not. I guess it should be sufficient to remove 

- https://github.com/galaxyproject/galaxy/blob/2e7459cc319789ba9a2612463eb9c87689e29c48/lib/galaxy/tool_util/deps/singularity_util.py#L82
- https://github.com/galaxyproject/galaxy/blob/2e7459cc319789ba9a2612463eb9c87689e29c48/lib/galaxy/tool_util/deps/singularity_util.py#L98

Extra volumes should be added via https://github.com/galaxyproject/galaxy/blob/2e7459cc319789ba9a2612463eb9c87689e29c48/lib/galaxy/config/sample/job_conf.sample.yml#L574 .. we can / should add a note to admins for the next release.

**Comment by rhpvorderman** [comment]
This is the commit where it was added. 
https://github.com/galaxyproject/galaxy/commit/658816264a8e84bacfdf83e9c9029338b2e1f85d

I also found the PR: https://github.com/galaxyproject/galaxy/pull/7125

I cannot find the rationale for adding the home directory though. That was probably something that fixed a test. There are no notes in the commit, no comments in the code, and I cannot find it mentioned in the PR.

**Comment by rhpvorderman** [comment]
Yes I agree with you @bernt-matthias . This should be beneficial for most users to do differently. For the small use case where home needs to be mounted, this can be done with the extra arguments param.

I also prefer to run with `--contain --ipc --pid` flags in addition to cleanenv. Especially `--contain` ensures that only directories explicitly mounted are mounted. 

Then I also found there were some problems with tools not being able to write to tmp and setting `<param id="tmp_dir>true</param>` fixes this. So the singularity default configuration can use some touching up, but I will do one thing at a time.

**Comment by rhpvorderman** [comment]
I see now the problem is a little more complicated. Older tools like fastqc require a galaxy home, while tools with profile 18.04 will provide a separate job home.
